include/fi_util.h

```diff
@@ -295,6 +295,32 @@ struct util_event {
 int fi_eq_create(struct fid_fabric *fabric, struct fi_eq_attr *attr,
 		 struct fid_eq **eq_fid, void *context);
 
+/*
+ * Buffer Pool
+ */
+
+struct util_buf_pool {
+	size_t size;
+	size_t num;
+	void *memory_region;
+	struct slist list;
+	fastlock_t lock;
+};
+
+struct util_buf {
+	struct slist_entry entry;
+	struct util_buf_pool *pool;
+	uint8_t data[0];
+};
```

Comment:
I would change this to:

`
union util_buf {
    struct slist_entry entry;
    uint8_t data[0]
};
`
and have the caller provider the pool as input into free.  This provides better control over the alignment of the data buffer, rather than it always being offset by the size of entry + *pooll